Widgets/Widgets/PictureOfTheDayWidget.swift

```diff
@@ -37,41 +37,69 @@ final class PictureOfTheDayData {
+            let moc = self.dataStore.viewContext
+            moc.perform {
+                guard let latest = moc.newestVisibleGroup(of: .pictureOfTheDay) else {
+                    completion(self.sampleEntry)
+                    return
+                }
+                self.assemblePictureEntryFromContentGroup(latest, completion: completion)
+            }
+        }
+
+    }
+
+    private func assemblePictureEntryFromContentGroup(_ contentGroup: WMFContentGroup, usingImageCache: Bool = false, completion: @escaping (PictureOfTheDayEntry) -> Void) {
+        guard let imageContent = contentGroup.contentPreview as? WMFFeedImage else {
+            completion(self.sampleEntry)
+            return
+        }
 
-		let sampleEntry = self.sampleEntry
-		let contentDate = contentGroup.date
-		let contentURL = contentGroup.url
+        let sampleEntry = self.sampleEntry
+        let contentDate = contentGroup.date
+        let contentURL = contentGroup.url
         let canonicalPageTitle = imageContent.canonicalPageTitle
-		let imageThumbnailURL = imageContent.imageThumbURL
-		let imageDescription = imageContent.imageDescription
-
-		guard !usingImageCache else {
-			if let cachedImage = ImageCacheController.shared?.memoryCachedImage(withURL: imageThumbnailURL) {
-				let entry = PictureOfTheDayEntry(date: Date(), contentDate: contentDate, contentURL: contentURL, imageURL: imageThumbnailURL, image: cachedImage.staticImage, imageDescription: imageDescription)
-				completion(entry)
-			} else {
-				completion(sampleEntry)
-			}
-			return
-		}
+        let imageThumbnailURL = imageContent.imageThumbURL
+        let imageDescription = imageContent.imageDescription
 
-		ImageCacheController.shared?.fetchImage(withURL: imageThumbnailURL, failure: { _ in
-			completion(sampleEntry)
-		}, success: { fetchedImage in
+        guard !usingImageCache else {
+            if let cachedImage = ImageCacheController.shared?.cachedImage(withURL: imageThumbnailURL) {
+                let entry = PictureOfTheDayEntry(date: Date(), contentDate: contentDate, contentURL: contentURL, imageURL: imageThumbnailURL, image: cachedImage.staticImage, imageDescription: imageDescription)
+                completion(entry)
+            } else {
+                completion(sampleEntry)
+            }
+            return
+        }
+
+        ImageCacheController.shared?.fetchImage(withURL: imageThumbnailURL, failure: { _ in
+            completion(sampleEntry)
+        }, success: { fetchedImage in
             self.fetchImageLicense(canonicalPageTitle: canonicalPageTitle) { license in
                 let entry = PictureOfTheDayEntry(date: Date(), contentDate: contentDate, contentURL: contentURL, imageURL: imageThumbnailURL, image: fetchedImage.image.staticImage, imageDescription: imageDescription, license: license)
                 completion(entry)
             }
-		})
-	}
-
-	// MARK: Private
+        })
+    }
```

Comment:
(it won't let me comment below so I'm doing it here) - I've been looking and I don't think the `imageInfoFetcher.fetchGalleryInfo` will run the completion on the main queue. We might need to wrap up that completion calls in a `DispatchQueue.main`

Reply:
Update - I went ahead and made this change. Seems to be testing okay.

Reply:
Nice catch – I got used to `fetchImage` on `ImageCacheController` wrapping the completion for me and forgot to do it myself here. Thanks for making the change.